<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mapping disease risk loci using varbvs • varbvs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">varbvs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/pcarbo/varbvs">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Mapping disease risk loci using varbvs</h1>
            
            <h4 class="date">2017-07-21</h4>
          </div>

    
    
<div class="contents">
<p>This vignettes demonstrates how to fit a Bayesian variable selection model using <strong>varbvs</strong> to identify genetic markers associated with Crohn’s disease risk. The data consist of 442,001 SNPs genotyped for 1,748 cases and 2,938 controls. Note that file <code>cd.RData</code> cannot be made publicly available due to data sharing restrictions, so this vignette is for viewing only.</p>
<p>Begin by loading a couple packages into the R environment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(lattice)
<span class="kw">library</span>(varbvs)</code></pre></div>
<p>Set the random number generator seed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1</span>)</code></pre></div>
<div id="load-the-genotype-and-phenotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-genotype-and-phenotype-data" class="anchor"></a>Load the genotype and phenotype data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">load</span>(<span class="st">"cd.RData"</span>)</code></pre></div>
</div>
<div id="fit-variational-approximation-to-posterior" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-variational-approximation-to-posterior" class="anchor"></a>Fit variational approximation to posterior</h2>
<p>Here we fit the fully-factorized variational approximation to the posterior distribution of the coefficients for a logistic regression model of a binary outcome (case-control status), with spike and slab priors on the coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">r &lt;-<span class="st"> </span><span class="kw">system.time</span>(fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/varbvs.html">varbvs</a></span>(X,<span class="ot">NULL</span>,y,<span class="st">"binomial"</span>,<span class="dt">logodds =</span> <span class="kw">seq</span>(-<span class="dv">6</span>,-<span class="dv">3</span>,<span class="fl">0.25</span>)))
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Model fitting took %0.2f minutes.</span><span class="ch">\n</span><span class="st">"</span>,r[<span class="st">"elapsed"</span>]/<span class="dv">60</span>))</code></pre></div>
<p>Compute “single-marker” posterior inclusion probabilities.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pip &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw"><a href="../reference/varbvsindep.html">varbvsindep</a></span>(fit,X,<span class="ot">NULL</span>,y)$alpha %*%<span class="st"> </span>fit$w)</code></pre></div>
</div>
<div id="save-the-results-to-a-file" class="section level2">
<h2 class="hasAnchor">
<a href="#save-the-results-to-a-file" class="anchor"></a>Save the results to a file</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(<span class="dt">list =</span> <span class="kw">c</span>(<span class="st">"fit"</span>,<span class="st">"map"</span>,<span class="st">"pip"</span>,<span class="st">"r"</span>),
     <span class="dt">file =</span> <span class="st">"varbvs.demo.cd.RData"</span>)</code></pre></div>
</div>
<div id="summarize-the-model-fitting" class="section level2">
<h2 class="hasAnchor">
<a href="#summarize-the-model-fitting" class="anchor"></a>Summarize the model fitting</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">print</span>(<span class="kw">summary</span>(fit,<span class="dt">nv =</span> <span class="dv">9</span>))</code></pre></div>
<p>Show two “genome-wide scans”, one using the posterior inclusion probabilities (PIPs) computed in the joint analysis of all variables, and one using the PIPs that ignore correlations between the variables. The latter is meant to look like a typical genome-wide “Manhattan” plot used to summarize the results of a genome-wide association study. Variables with <code>PIP &gt; 0.5</code> are highlighted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">i &lt;-<span class="st"> </span><span class="kw">which</span>(fit$pip &gt;<span class="st"> </span><span class="fl">0.5</span>)
var.labels &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="kw">round</span>(map$pos[i]/<span class="fl">1e6</span>,<span class="dt">digits =</span> <span class="dv">2</span>),<span class="st">"Mb"</span>)
<span class="kw">print</span>(<span class="kw">plot</span>(fit,<span class="dt">groups =</span> map$chr,<span class="dt">vars =</span> i,<span class="dt">var.labels =</span> var.labels,<span class="dt">gap =</span> <span class="dv">7500</span>,
           <span class="dt">ylab =</span> <span class="st">"posterior prob."</span>),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)
<span class="kw">print</span>(<span class="kw">plot</span>(fit,<span class="dt">groups =</span> map$chr,<span class="dt">score =</span> <span class="kw">log10</span>(pip +<span class="st"> </span><span class="fl">0.001</span>),<span class="dt">vars =</span> i,
           <span class="dt">var.labels =</span> var.labels,<span class="dt">gap =</span> <span class="dv">7500</span>,<span class="dt">ylab =</span> <span class="st">"log10 posterior prob."</span>),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-the-genotype-and-phenotype-data">Load the genotype and phenotype data</a></li>
      <li><a href="#fit-variational-approximation-to-posterior">Fit variational approximation to posterior</a></li>
      <li><a href="#save-the-results-to-a-file">Save the results to a file</a></li>
      <li><a href="#summarize-the-model-fitting">Summarize the model fitting</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Peter Carbonetto, Matthew Stephens, David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
