<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparison of glmnet and varbvs in Leukemia data set • varbvs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">varbvs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/pcarbo/varbvs">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Comparison of glmnet and varbvs in Leukemia data set</h1>
            
            <h4 class="date">2017-07-21</h4>
          </div>

    
    
<div class="contents">
<p>This vignette demonstrates application of <strong>glmnet</strong> and <strong>varbvs</strong> to the Leukemia data set. The main aim of this script is to illustrate some of the different properties of Bayesian variable selection and penalized sparse regression (as implemented by <strong>varbvs</strong> and <strong>glmnet</strong>, respectively).</p>
<p>We use the preprocessed data of Dettling (2004) retrieved from the supplementary materials accompanying Friedman <em>et al</em> (2010). The data are represented as a 72 x 3,571 matrix of gene expression values (variable <code>X</code>), and a vector of 72 binary disease outcomes (variable <code>y</code>).</p>
<div id="vignette-parameters" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette-parameters" class="anchor"></a>Vignette parameters</h2>
<p>Begin by loading these packages into your R environment.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(lattice)
<span class="kw">library</span>(latticeExtra)
<span class="kw">library</span>(glmnet)
<span class="kw">library</span>(varbvs)</code></pre></div>
<p>Specify settings for the glmnet analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nfolds &lt;-<span class="st"> </span><span class="dv">20</span>                    <span class="co"># Number of cross-validation folds.</span>
alpha  &lt;-<span class="st"> </span><span class="fl">0.95</span>                  <span class="co"># Elastic net mixing parameter.</span>
lambda &lt;-<span class="st"> </span><span class="dv">10</span>^(<span class="kw">seq</span>(<span class="dv">0</span>,-<span class="dv">2</span>,-<span class="fl">0.05</span>))  <span class="co"># Lambda sequence.</span></code></pre></div>
</div>
<div id="load-the-leukemia-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-leukemia-data" class="anchor"></a>Load the Leukemia data</h2>
<p>Also set the random number generator seed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(leukemia)
X &lt;-<span class="st"> </span>leukemia$x
y &lt;-<span class="st"> </span>leukemia$y
<span class="kw">set.seed</span>(<span class="dv">1</span>)</code></pre></div>
</div>
<div id="fit-elastic-net-model-to-data" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-elastic-net-model-to-data" class="anchor"></a>Fit elastic net model to data</h2>
<p>Here, we also run 20-fold cross-validation to select the largest setting of the L1-penalty strength (<em>lambda</em>) that is within 1 standard error of the minimum classification error.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># This is the model fitting step.</span>
r &lt;-<span class="st"> </span><span class="kw">system.time</span>(fit.glmnet &lt;-
<span class="st">       </span><span class="kw">glmnet</span>(X,y,<span class="dt">family =</span> <span class="st">"binomial"</span>,<span class="dt">lambda =</span> lambda,<span class="dt">alpha =</span> alpha))
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Model fitting took %0.2f seconds.</span><span class="ch">\n</span><span class="st">"</span>,r[<span class="st">"elapsed"</span>]))
<span class="co"># Model fitting took 0.06 seconds.</span>

<span class="co"># This is the cross-validation step.</span>
r &lt;-<span class="st"> </span><span class="kw">system.time</span>(out.cv.glmnet &lt;-
<span class="st">       </span><span class="kw">cv.glmnet</span>(X,y,<span class="dt">family =</span> <span class="st">"binomial"</span>,<span class="dt">type.measure =</span> <span class="st">"class"</span>,
                 <span class="dt">alpha =</span> alpha,<span class="dt">nfolds =</span> nfolds,<span class="dt">lambda =</span> lambda))
lambda &lt;-<span class="st"> </span>out.cv.glmnet$lambda
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Cross-validation took %0.2f seconds.</span><span class="ch">\n</span><span class="st">"</span>,r[<span class="st">"elapsed"</span>]))
<span class="co"># Cross-validation took 2.24 seconds.</span>

<span class="co"># Choose the largest value of lambda that is within 1 standard error</span>
<span class="co"># of the smallest misclassification error.</span>
lambda.opt &lt;-<span class="st"> </span>out.cv.glmnet$lambda.1se</code></pre></div>
</div>
<div id="evaluate-the-glmnet-predictions" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluate-the-glmnet-predictions" class="anchor"></a>Evaluate the glmnet predictions</h2>
<p>Compute estimates of the disease outcome using the fitted model, and compare against the observed values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="st">"classification results with lambda = "</span>,lambda.opt,<span class="st">":</span><span class="ch">\n</span><span class="st">"</span>,<span class="dt">sep=</span><span class="st">""</span>)
y.glmnet &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">predict</span>(fit.glmnet,X,<span class="dt">s =</span> lambda.opt,<span class="dt">type =</span> <span class="st">"class"</span>))
<span class="kw">print</span>(<span class="kw">table</span>(<span class="dt">true =</span> <span class="kw">factor</span>(y),<span class="dt">pred =</span> <span class="kw">factor</span>(y.glmnet)))
<span class="co"># classification results with lambda = 0.2238721:</span>
<span class="co">#     pred</span>
<span class="co"># true  0  1</span>
<span class="co">#    0 47  0</span>
<span class="co">#    1  4 21</span></code></pre></div>
</div>
<div id="visualize-results-of-glmnet-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-results-of-glmnet-analysis" class="anchor"></a>Visualize results of glmnet analysis</h2>
<p>The first plot shows the evolution of regression coefficients at different settings of <em>lambda</em>. (Note that the intercept is not shown.) Only the curves for the variables that are selected at the optimal setting of <em>lambda</em> (“lambda.opt”“) are labeled.</p>
<p>The second plot shows the classification error at different settings of <em>lambda</em>.</p>
<p>The third plot shows the number of nonzero regression coefficients at different settings of <em>lambda</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">trellis.par.set</span>(<span class="dt">par.xlab.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.85</span>),
                <span class="dt">par.ylab.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.85</span>),
                <span class="dt">axis.text     =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.75</span>))

<span class="co"># Choose the largest value of lambda that is within 1 standard error</span>
<span class="co"># of the smallest misclassification error.</span>
lambda.opt &lt;-<span class="st"> </span>out.cv.glmnet$lambda.1se

<span class="co"># Plot regression coefficients.</span>
lambda   &lt;-<span class="st"> </span>fit.glmnet$lambda
vars     &lt;-<span class="st"> </span><span class="kw">setdiff</span>(<span class="kw">which</span>(<span class="kw">rowSums</span>(<span class="kw">abs</span>(<span class="kw">coef</span>(fit.glmnet))) &gt;<span class="st"> </span><span class="dv">0</span>),<span class="dv">1</span>)
n        &lt;-<span class="st"> </span><span class="kw">length</span>(vars)
b        &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">t</span>(<span class="kw">coef</span>(fit.glmnet)[vars,]))
i        &lt;-<span class="st"> </span><span class="kw">coef</span>(fit.glmnet,<span class="dt">s =</span> lambda.opt)
i        &lt;-<span class="st"> </span><span class="kw">rownames</span>(i)[<span class="kw">which</span>(i !=<span class="st"> </span><span class="dv">0</span>)]
i        &lt;-<span class="st"> </span>i[-<span class="dv">1</span>]
vars.opt &lt;-<span class="st"> </span><span class="kw">colnames</span>(b)
vars.opt[!<span class="kw">is.element</span>(vars.opt,i)] &lt;-<span class="st"> ""</span>
vars.opt &lt;-<span class="st"> </span><span class="kw">substring</span>(vars.opt,<span class="dv">2</span>)
lab  &lt;-<span class="st"> </span><span class="kw">expression</span>(<span class="st">"more complex"</span> %&lt;-%<span class="st"> </span><span class="kw">paste</span>(log[<span class="dv">10</span>],lambda) %-&gt;%<span class="st"> </span>
<span class="st">                   "less complex"</span>)
r    &lt;-<span class="st"> </span><span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> b[,<span class="dv">1</span>]),<span class="dt">type =</span> <span class="st">"l"</span>,
               <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">xlab =</span> lab,<span class="dt">ylab =</span> <span class="st">"regression coefficient"</span>,
               <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">2.35</span>,<span class="fl">0.1</span>)),
                             <span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.8</span>,<span class="fl">1.2</span>))),
               <span class="dt">panel =</span> function(x, y, ...) {
                 <span class="kw">panel.xyplot</span>(x,y,...);
                 <span class="kw">panel.abline</span>(<span class="dt">v =</span> <span class="kw">log10</span>(lambda.opt),<span class="dt">col =</span> <span class="st">"orangered"</span>,
                              <span class="dt">lwd =</span> <span class="dv">2</span>,<span class="dt">lty =</span> <span class="st">"dotted"</span>);
                 <span class="kw">ltext</span>(<span class="dt">x =</span> -<span class="dv">2</span>,<span class="dt">y =</span> b[<span class="kw">nrow</span>(b),],<span class="dt">labels =</span> vars.opt,<span class="dt">pos =</span> <span class="dv">2</span>,
                       <span class="dt">offset =</span> <span class="fl">0.5</span>,<span class="dt">cex =</span> <span class="fl">0.5</span>);
               })
for (i in <span class="dv">2</span>:n)
  r &lt;-<span class="st"> </span>r +<span class="st"> </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> b[,i]),
                           <span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"blue"</span>))
<span class="kw">print</span>(r,<span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)

<span class="co"># Plot classification error.</span>
Y       &lt;-<span class="st"> </span><span class="kw">predict</span>(fit.glmnet,X,<span class="dt">type =</span> <span class="st">"class"</span>)
<span class="kw">mode</span>(Y) &lt;-<span class="st"> "numeric"</span>
<span class="kw">print</span>(<span class="kw">with</span>(out.cv.glmnet,
           <span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> cvm),<span class="dt">type =</span> <span class="st">"l"</span>,
                  <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">xlab =</span> lab,
                  <span class="dt">ylab =</span> <span class="st">"20-fold cross-validation </span><span class="ch">\n</span><span class="st"> classification error"</span>,
                  <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.02</span>,<span class="fl">0.45</span>))),
                  <span class="dt">panel =</span> function(x, y, ...) {
                    <span class="kw">panel.xyplot</span>(x,y,...);
                    <span class="kw">panel.abline</span>(<span class="dt">v =</span> <span class="kw">log10</span>(lambda.opt),<span class="dt">col =</span> <span class="st">"orangered"</span>,
                                 <span class="dt">lwd =</span> <span class="dv">2</span>,<span class="dt">lty =</span> <span class="st">"dotted"</span>);
                  }) +
<span class="st">           </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> cvm),
                           <span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">cex =</span> <span class="fl">0.6</span>,<span class="dt">col =</span> <span class="st">"blue"</span>)) +
<span class="st">           </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> cvup),
                           <span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">lty =</span> <span class="st">"solid"</span>)) +
<span class="st">           </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> cvlo),
                           <span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">lty =</span> <span class="st">"solid"</span>)) +
<span class="st">           </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),
                                            <span class="dt">y =</span> <span class="kw">colMeans</span>(<span class="kw">abs</span>(Y -<span class="st"> </span>y))),
                           <span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"darkorange"</span>,<span class="dt">lwd =</span> <span class="dv">2</span>,
                           <span class="dt">lty =</span> <span class="st">"solid"</span>))),
           <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)

<span class="co"># Plot number of non-zero regression coefficients.</span>
<span class="kw">print</span>(<span class="kw">with</span>(out.cv.glmnet,
           <span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> nzero),<span class="dt">type =</span> <span class="st">"l"</span>,
                  <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">xlab =</span> lab,
                  <span class="dt">ylab =</span> <span class="st">"number of non-zero </span><span class="ch">\n</span><span class="st"> coefficients"</span>,
                  <span class="dt">panel =</span> function(x, y, ...) {
                    <span class="kw">panel.xyplot</span>(x,y,...)
                    <span class="kw">panel.abline</span>(<span class="dt">v =</span> <span class="kw">log10</span>(lambda.opt),<span class="dt">col =</span> <span class="st">"orangered"</span>,
                                 <span class="dt">lwd =</span> <span class="dv">2</span>,<span class="dt">lty =</span> <span class="st">"dotted"</span>)
                  }) +
<span class="st">           </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">log10</span>(lambda),<span class="dt">y =</span> nzero),
                           <span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">cex =</span> <span class="fl">0.6</span>,<span class="dt">col =</span> <span class="st">"blue"</span>))),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="leukemia_files/figure-html/unnamed-chunk-7-1.png" alt="&amp;nbsp;" width="828"><p class="caption">
 
</p>
</div>
</div>
<div id="fit-variational-approximation-to-posterior" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-variational-approximation-to-posterior" class="anchor"></a>Fit variational approximation to posterior</h2>
<p>Fit the fully-factorized variational approximation to the posterior distribution of the coefficients for a logistic regression model of the binary outcome (the type of leukemia), with spike-and-slab priors on the coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">r &lt;-<span class="st"> </span><span class="kw">system.time</span>(fit.varbvs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/varbvs.html">varbvs</a></span>(X,<span class="ot">NULL</span>,y,<span class="st">"binomial"</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>))
<span class="kw">cat</span>(<span class="kw">sprintf</span>(<span class="st">"Model fitting took %0.2f seconds.</span><span class="ch">\n</span><span class="st">"</span>,r[<span class="st">"elapsed"</span>]))
<span class="co"># Model fitting took 31.50 seconds.</span></code></pre></div>
</div>
<div id="evaluate-the-varbvs-predictions" class="section level2">
<h2 class="hasAnchor">
<a href="#evaluate-the-varbvs-predictions" class="anchor"></a>Evaluate the varbvs predictions</h2>
<p>Compute estimates of the disease outcome using the fitted model, and compare against the observed values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y.varbvs &lt;-<span class="st"> </span><span class="kw">predict</span>(fit.varbvs,X)
<span class="kw">print</span>(<span class="kw">table</span>(<span class="dt">true =</span> <span class="kw">factor</span>(y),<span class="dt">pred =</span> <span class="kw">factor</span>(y.varbvs)))
<span class="co">#     pred</span>
<span class="co"># true  0  1</span>
<span class="co">#    0 44  3</span>
<span class="co">#    1  3 22</span></code></pre></div>
</div>
<div id="visualize-results-of-varbvs-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-results-of-varbvs-analysis" class="anchor"></a>Visualize results of varbvs analysis</h2>
<p>The first plot shows the classification error at each setting of the prior log-odds.</p>
<p>The second plot shows the evolution of the posterior mean regression coefficients (the beta’s) at different settings of the prior log-odds, for the top 6 variables ranked by posterior inclusion probability (averaged over settings of the hyperparameters).</p>
<p>The top-ranked variable (by posterior inclusion probability) has a much larger coefficient than all the others, so it is shown in a separate plot.</p>
<p>The third plot shows the (approximate) probability density of the prior log-odds parameter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">trellis.par.set</span>(<span class="dt">par.xlab.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.85</span>),
                <span class="dt">par.ylab.text =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.85</span>),
                <span class="dt">axis.text     =</span> <span class="kw">list</span>(<span class="dt">cex =</span> <span class="fl">0.75</span>))

<span class="co"># Get the normalized importance weights.</span>
w &lt;-<span class="st"> </span>fit.varbvs$w

<span class="co"># Plot classification error at each hyperparameter setting.</span>
sigmoid10 &lt;-<span class="st"> </span>function (x)
  <span class="dv">1</span>/(<span class="dv">1</span> +<span class="st"> </span><span class="dv">10</span>^(-x))
logodds &lt;-<span class="st"> </span>fit.varbvs$logodds
log10q  &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">sigmoid10</span>(logodds))
m       &lt;-<span class="st"> </span><span class="kw">length</span>(logodds)
err     &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>,m)
for (i in <span class="dv">1</span>:m) {
  r      &lt;-<span class="st"> </span><span class="kw">subset</span>(fit.varbvs,logodds ==<span class="st"> </span>logodds[i])
  ypred  &lt;-<span class="st"> </span><span class="kw">predict</span>(r,X)
  err[i] &lt;-<span class="st"> </span><span class="kw">mean</span>(y !=<span class="st"> </span>ypred)
}
lab &lt;-<span class="st"> </span><span class="kw">expression</span>(<span class="st">"more complex"</span> %&lt;-%<span class="st"> </span><span class="kw">paste</span>(log[<span class="dv">10</span>],pi) %-&gt;%<span class="st"> "less complex"</span>)
<span class="kw">print</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> err),<span class="dt">type =</span> <span class="st">"l"</span>,
             <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">xlab =</span> lab,<span class="dt">ylab =</span> <span class="st">"classification error"</span>,
             <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.9</span>,-<span class="fl">3.65</span>)))) +
<span class="st">      </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> err),
                      <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">cex =</span> <span class="fl">0.65</span>)),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)

<span class="co"># Plot expected number of included variables at each hyperparameter</span>
<span class="co"># setting.</span>
r &lt;-<span class="st"> </span><span class="kw">colSums</span>(fit.varbvs$alpha)
<span class="kw">print</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> r),<span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"blue"</span>,
             <span class="dt">xlab =</span> lab,<span class="dt">ylab =</span> <span class="st">"expected number of</span><span class="ch">\n</span><span class="st">included variables"</span>,
             <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.9</span>,-<span class="fl">3.65</span>)),
                           <span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">log =</span> <span class="dv">10</span>,<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">10</span>,<span class="dv">100</span>)))) +
<span class="st">      </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> r),
                      <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">cex =</span> <span class="fl">0.65</span>,
                      <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.9</span>,-<span class="fl">3.65</span>)),
                                    <span class="dt">y =</span> <span class="kw">list</span>(<span class="dt">log =</span> <span class="dv">10</span>)))),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>),<span class="dt">more =</span> <span class="ot">TRUE</span>)

<span class="co"># Plot density of prior inclusion probability hyperparameter.</span>
<span class="kw">print</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> w),<span class="dt">type =</span> <span class="st">"l"</span>,<span class="dt">col =</span> <span class="st">"blue"</span>,
             <span class="dt">xlab =</span> lab,
             <span class="dt">ylab =</span> <span class="kw">expression</span>(<span class="kw">paste</span>(<span class="st">"posterior probability of "</span>,pi)),
             <span class="dt">scales =</span> <span class="kw">list</span>(<span class="dt">x =</span> <span class="kw">list</span>(<span class="dt">limits =</span> <span class="kw">c</span>(-<span class="fl">0.9</span>,-<span class="fl">3.65</span>)))) +
<span class="st">      </span><span class="kw">as.layer</span>(<span class="kw">xyplot</span>(y ~<span class="st"> </span>x,<span class="kw">data.frame</span>(<span class="dt">x =</span> log10q,<span class="dt">y =</span> w),
                      <span class="dt">col =</span> <span class="st">"blue"</span>,<span class="dt">pch =</span> <span class="dv">20</span>,<span class="dt">cex =</span> <span class="fl">0.65</span>)),
      <span class="dt">split =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>),<span class="dt">more =</span> <span class="ot">FALSE</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="leukemia_files/figure-html/unnamed-chunk-10-1.png" alt="&amp;nbsp;" width="828"><p class="caption">
 
</p>
</div>
<div id="references" class="section level3">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
<p>Dettling, M. (2004). BagBoosting for tumor classification with gene expression data. <em>Bioinformatics</em> <strong>20</strong>, 3583–3593.</p>
<p>Friedman, J., Hastie, T., Tibshirani, R. (2010) Regularization paths for generalized linear models via coordinate descent. <em>Journal of Statistical Software</em> <strong>33</strong>, 1–22.</p>
</div>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h2>
<p>This is the version of R and the packages that were used to generate these results.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co"># R version 3.3.2 (2016-10-31)</span>
<span class="co"># Platform: x86_64-apple-darwin13.4.0 (64-bit)</span>
<span class="co"># Running under: macOS Sierra 10.12.5</span>
<span class="co"># </span>
<span class="co"># locale:</span>
<span class="co"># [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co"># </span>
<span class="co"># attached base packages:</span>
<span class="co"># [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co"># </span>
<span class="co"># other attached packages:</span>
<span class="co"># [1] varbvs_2.3-1        glmnet_2.0-5        foreach_1.4.3      </span>
<span class="co"># [4] Matrix_1.2-8        latticeExtra_0.6-28 RColorBrewer_1.1-2 </span>
<span class="co"># [7] lattice_0.20-34    </span>
<span class="co"># </span>
<span class="co"># loaded via a namespace (and not attached):</span>
<span class="co">#  [1] Rcpp_0.12.12     codetools_0.2-15 digest_0.6.12    rprojroot_1.2   </span>
<span class="co">#  [5] grid_3.3.2       backports_1.0.5  magrittr_1.5     evaluate_0.10.1 </span>
<span class="co">#  [9] highr_0.6        stringi_1.1.2    rmarkdown_1.6    iterators_1.0.8 </span>
<span class="co"># [13] tools_3.3.2      stringr_1.2.0    yaml_2.1.14      htmltools_0.3.6 </span>
<span class="co"># [17] knitr_1.16</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#vignette-parameters">Vignette parameters</a></li>
      <li><a href="#load-the-leukemia-data">Load the Leukemia data</a></li>
      <li><a href="#fit-elastic-net-model-to-data">Fit elastic net model to data</a></li>
      <li><a href="#evaluate-the-glmnet-predictions">Evaluate the glmnet predictions</a></li>
      <li><a href="#visualize-results-of-glmnet-analysis">Visualize results of glmnet analysis</a></li>
      <li><a href="#fit-variational-approximation-to-posterior">Fit variational approximation to posterior</a></li>
      <li><a href="#evaluate-the-varbvs-predictions">Evaluate the varbvs predictions</a></li>
      <li><a href="#visualize-results-of-varbvs-analysis">Visualize results of varbvs analysis</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Peter Carbonetto, Matthew Stephens, David Gerard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
